File: regions/system/library/dart/config/test/vault_test.dart
import 'package:test/test.dart';
import 'package:k1s0_config/config.dart';

Config minimalConfig({
  DatabaseConfig? database,
  RedisConfig? redis,
  RedisConfig? redisSession,
  KafkaConfig? kafka,
  AuthConfig? auth,
}) {
  return Config(
    app: AppConfig(
        name: 'test', version: '1.0', tier: 'system', environment: 'dev'),
    server: ServerConfig(host: '0.0.0.0', port: 8080),
    database: database,
    redis: redis,
    redisSession: redisSession,
    kafka: kafka,
    observability: ObservabilityConfig(
      log: LogConfig(level: 'info', format: 'json'),
      trace: TraceConfig(enabled: false),
      metrics: MetricsConfig(enabled: false),
    ),
    auth: auth ??
        AuthConfig(jwt: JwtConfig(issuer: 'http://localhost', audience: 'test')),
  );
}

void main() {
  group('mergeVaultSecrets', () {
    test('should merge database.password', () {
      final cfg = minimalConfig(
        database: DatabaseConfig(
          host: 'localhost',
          port: 5432,
          name: 'test_db',
          user: 'app',
          password: 'old',
        ),
      );
      mergeVaultSecrets(cfg, {'database.password': 'vault-db-pass'});
      expect(cfg.database!.password, 'vault-db-pass');
    });

    test('should merge redis.password', () {
      final cfg = minimalConfig(
        redis: RedisConfig(host: 'localhost', port: 6379, password: 'old'),
      );
      mergeVaultSecrets(cfg, {'redis.password': 'vault-redis-pass'});
      expect(cfg.redis!.password, 'vault-redis-pass');
    });

    test('should merge kafka.sasl credentials', () {
      final cfg = minimalConfig(
        kafka: KafkaConfig(
          brokers: ['localhost:9092'],
          consumerGroup: 'test.default',
          securityProtocol: 'SASL_SSL',
          sasl: KafkaSaslConfig(
            mechanism: 'SCRAM-SHA-512',
            username: '',
            password: '',
          ),
          topics: KafkaTopics(publish: [], subscribe: []),
        ),
      );
      mergeVaultSecrets(cfg, {
        'kafka.sasl.username': 'vault-kafka-user',
        'kafka.sasl.password': 'vault-kafka-pass',
      });
      expect(cfg.kafka!.sasl!.username, 'vault-kafka-user');
      expect(cfg.kafka!.sasl!.password, 'vault-kafka-pass');
    });

    test('should merge redis_session.password', () {
      final cfg = minimalConfig(
        redisSession:
            RedisConfig(host: 'localhost', port: 6380, password: ''),
      );
      mergeVaultSecrets(
          cfg, {'redis_session.password': 'vault-session-pass'});
      expect(cfg.redisSession!.password, 'vault-session-pass');
    });

    test('should merge oidc.client_secret', () {
      final cfg = minimalConfig(
        auth: AuthConfig(
          jwt: JwtConfig(issuer: 'http://localhost', audience: 'test'),
          oidc: OidcConfig(
            discoveryUrl: 'http://localhost/.well-known',
            clientId: 'test',
            redirectUri: 'http://localhost/callback',
            scopes: ['openid'],
            jwksUri: 'http://localhost/jwks',
          ),
        ),
      );
      mergeVaultSecrets(
          cfg, {'auth.oidc.client_secret': 'vault-oidc-secret'});
      expect(cfg.auth.oidc!.clientSecret, 'vault-oidc-secret');
    });

    test('should not change config when secrets are empty', () {
      final cfg = minimalConfig(
        database: DatabaseConfig(
          host: 'localhost',
          port: 5432,
          name: 'test_db',
          user: 'app',
          password: 'original',
        ),
        redis:
            RedisConfig(host: 'localhost', port: 6379, password: 'original'),
      );
      mergeVaultSecrets(cfg, {});
      expect(cfg.database!.password, 'original');
      expect(cfg.redis!.password, 'original');
    });

    test('should handle nil sections safely', () {
      final cfg = minimalConfig();
      // All optional sections are null - should not throw
      mergeVaultSecrets(cfg, {
        'database.password': 'secret',
        'redis.password': 'secret',
        'kafka.sasl.username': 'user',
        'kafka.sasl.password': 'pass',
        'redis_session.password': 'secret',
        'auth.oidc.client_secret': 'secret',
      });
      expect(cfg.database, isNull);
      expect(cfg.redis, isNull);
      expect(cfg.kafka, isNull);
      expect(cfg.redisSession, isNull);
      expect(cfg.auth.oidc, isNull);
    });

    test('should merge only existing partial secrets', () {
      final cfg = minimalConfig(
        database: DatabaseConfig(
          host: 'localhost',
          port: 5432,
          name: 'test_db',
          user: 'app',
          password: 'old-db',
        ),
        redis:
            RedisConfig(host: 'localhost', port: 6379, password: 'old-redis'),
        auth: AuthConfig(
          jwt: JwtConfig(issuer: 'http://localhost', audience: 'test'),
          oidc: OidcConfig(
            discoveryUrl: 'http://localhost/.well-known',
            clientId: 'test',
            clientSecret: 'old-oidc',
            redirectUri: 'http://localhost/callback',
            scopes: ['openid'],
            jwksUri: 'http://localhost/jwks',
          ),
        ),
      );
      // Only database.password is provided
      mergeVaultSecrets(cfg, {'database.password': 'new-db'});
      expect(cfg.database!.password, 'new-db');
      expect(cfg.redis!.password, 'old-redis');
      expect(cfg.auth.oidc!.clientSecret, 'old-oidc');
    });
  });
}

File: regions/system/library/typescript/config/__tests__/vault.test.ts
import { describe, it, expect } from 'vitest';
import { mergeVaultSecrets } from '../src/index.js';
import type { Config } from '../src/index.js';

function minimalConfig(overrides?: Partial<Config>): Config {
  return {
    app: {
      name: 'test',
      version: '1.0',
      tier: 'system',
      environment: 'dev',
    },
    server: { host: '0.0.0.0', port: 8080 },
    observability: {
      log: { level: 'info', format: 'json' },
      trace: { enabled: false },
      metrics: { enabled: false },
    },
    auth: {
      jwt: { issuer: 'http://localhost', audience: 'test' },
    },
    ...overrides,
  } as Config;
}

describe('mergeVaultSecrets', () => {
  it('should merge database.password', () => {
    const cfg = minimalConfig({
      database: {
        host: 'localhost',
        port: 5432,
        name: 'test_db',
        user: 'app',
        password: 'old',
      },
    });
    const merged = mergeVaultSecrets(cfg, {
      'database.password': 'vault-db-pass',
    });
    expect(merged.database?.password).toBe('vault-db-pass');
  });

  it('should merge redis.password', () => {
    const cfg = minimalConfig({
      redis: { host: 'localhost', port: 6379, password: 'old' },
    });
    const merged = mergeVaultSecrets(cfg, {
      'redis.password': 'vault-redis-pass',
    });
    expect(merged.redis?.password).toBe('vault-redis-pass');
  });

  it('should merge kafka.sasl credentials', () => {
    const cfg = minimalConfig({
      kafka: {
        brokers: ['localhost:9092'],
        consumer_group: 'test.default',
        security_protocol: 'SASL_SSL',
        sasl: {
          mechanism: 'SCRAM-SHA-512',
          username: '',
          password: '',
        },
        topics: { publish: [], subscribe: [] },
      },
    });
    const merged = mergeVaultSecrets(cfg, {
      'kafka.sasl.username': 'vault-kafka-user',
      'kafka.sasl.password': 'vault-kafka-pass',
    });
    expect(merged.kafka?.sasl?.username).toBe('vault-kafka-user');
    expect(merged.kafka?.sasl?.password).toBe('vault-kafka-pass');
  });

  it('should merge redis_session.password', () => {
    const cfg = minimalConfig({
      redis_session: { host: 'localhost', port: 6380, password: '' },
    });
    const merged = mergeVaultSecrets(cfg, {
      'redis_session.password': 'vault-session-pass',
    });
    expect(merged.redis_session?.password).toBe('vault-session-pass');
  });

  it('should merge oidc.client_secret', () => {
    const cfg = minimalConfig({
      auth: {
        jwt: { issuer: 'http://localhost', audience: 'test' },
        oidc: {
          discovery_url: 'http://localhost/.well-known',
          client_id: 'test',
          redirect_uri: 'http://localhost/callback',
          scopes: ['openid'],
          jwks_uri: 'http://localhost/jwks',
        },
      },
    });
    const merged = mergeVaultSecrets(cfg, {
      'auth.oidc.client_secret': 'vault-oidc-secret',
    });
    expect(merged.auth.oidc?.client_secret).toBe('vault-oidc-secret');
  });

  it('should not change config when secrets are empty', () => {
    const cfg = minimalConfig({
      database: {
        host: 'localhost',
        port: 5432,
        name: 'test_db',
        user: 'app',
        password: 'original',
      },
      redis: { host: 'localhost', port: 6379, password: 'original' },
    });
    const merged = mergeVaultSecrets(cfg, {});
    expect(merged.database?.password).toBe('original');
    expect(merged.redis?.password).toBe('original');
  });

  it('should handle nil/undefined sections safely', () => {
    const cfg = minimalConfig();
    // No database, redis, kafka, redis_session, or oidc sections
    const merged = mergeVaultSecrets(cfg, {
      'database.password': 'secret',
      'redis.password': 'secret',
      'kafka.sasl.username': 'user',
      'kafka.sasl.password': 'pass',
      'redis_session.password': 'secret',
      'auth.oidc.client_secret': 'secret',
    });
    expect(merged.database).toBeUndefined();
    expect(merged.redis).toBeUndefined();
    expect(merged.kafka).toBeUndefined();
    expect(merged.redis_session).toBeUndefined();
    expect(merged.auth.oidc).toBeUndefined();
  });

  it('should merge only existing partial secrets', () => {
    const cfg = minimalConfig({
      database: {
        host: 'localhost',
        port: 5432,
        name: 'test_db',
        user: 'app',
        password: 'old-db',
      },
      redis: { host: 'localhost', port: 6379, password: 'old-redis' },
      auth: {
        jwt: { issuer: 'http://localhost', audience: 'test' },
        oidc: {
          discovery_url: 'http://localhost/.well-known',
          client_id: 'test',
          client_secret: 'old-oidc',
          redirect_uri: 'http://localhost/callback',
          scopes: ['openid'],
          jwks_uri: 'http://localhost/jwks',
        },
      },
    });
    // Only database.password is provided
    const merged = mergeVaultSecrets(cfg, {
      'database.password': 'new-db',
    });
    expect(merged.database?.password).toBe('new-db');
    expect(merged.redis?.password).toBe('old-redis');
    expect(merged.auth.oidc?.client_secret).toBe('old-oidc');
  });
});
